service/utility.go

```diff
@@ -478,3 +479,106 @@ func isStateAvailable() bool {
+	if queryParams == nil {
+		return nil, errNilQueryTransaction
+	}
+
+	lastNames, firstNames := extractPublishersFields(queryParams.GetPublishers())
+	cities, states, provinces, countries := extractStudySitesFields(queryParams.GetStudySites())
+
+	pipeline := bson.NewArray(
+		bson.VC.DocumentFromElements(
+			bson.EC.SubDocumentFromElements(
+				"$match",
+				bson.EC.SubDocumentFromElements("publisherName.lastName",
+					buildArrayFromElements(lastNames)),
+				bson.EC.SubDocumentFromElements("publisherName.firstName",
+					buildArrayFromElements(firstNames)),
+
+				bson.EC.SubDocumentFromElements("studySite.city",
+					buildArrayFromElements(cities)),
+				bson.EC.SubDocumentFromElements("studySite.state",
+					buildArrayFromElements(states)),
+				bson.EC.SubDocumentFromElements("studySite.province",
+					buildArrayFromElements(provinces)),
+				bson.EC.SubDocumentFromElements("studySite.country",
+					buildArrayFromElements(countries)),
+
+				bson.EC.SubDocumentFromElements("callTypeName",
+					buildArrayFromElements(queryParams.GetCallTypes())),
+
+				bson.EC.SubDocumentFromElements("groundType",
+					buildArrayFromElements(queryParams.GetGroundTypes())),
+
+				bson.EC.SubDocumentFromElements("sensorType",
+					buildArrayFromElements(queryParams.GetSensorTypes())),
+
+				bson.EC.SubDocumentFromElements("sensorName",
+					buildArrayFromElements(queryParams.GetSensorNames())),
+			),
+		),
+	)
+
+	return pipeline, nil
+}
+
+func buildArrayFromElements(elems []string) *bson.Element {
+	if elems == nil || len(elems) == 0 {
+		return bson.EC.ArrayFromElements("$in", bson.VC.Regex(".*", ""))
+	}
+	elemVals := make([]*bson.Value, len(elems))
+	for i := 0; i < len(elems); i++ {
+		elemVals[i] = bson.VC.String(elems[i])
+	}
+
+	return bson.EC.ArrayFromElements("$in", elemVals...)
+}
+
+func extractPublishersFields(publishers []*pb.Publisher) ([]string, []string) {
+	if publishers == nil || len(publishers) == 0 {
+		return []string{}, []string{}
+	}
+
+	lastNames := make([]string, len(publishers))
+	firstNames := make([]string, len(publishers))
+
+	for i := 0; i < len(publishers); i++ {
+		lastNames[i] = publishers[i].GetLastName()
+		firstNames[i] = publishers[i].GetFirstName()
+	}
+
+	return lastNames, firstNames
+}
+
+func extractStudySitesFields(studySites []*pb.StudySite) ([]string, []string, []string, []string) {
+	if studySites == nil || len(studySites) == 0 {
+		return []string{}, []string{}, []string{}, []string{}
+	}
+
+	cities := make([]string, len(studySites))
+	states := make([]string, 0)
+	provinces := make([]string, 0)
+	countries := make([]string, len(studySites))
+
+	for i := 0; i < len(studySites); i++ {
+		cities[i] = studySites[i].GetCity()
+
+		tempState := strings.TrimSpace(studySites[i].GetState())
+		if tempState != "" {
+			states = append(states, tempState)
+		}
+
+		tempProvince := strings.TrimSpace(studySites[i].GetProvince())
+		if tempState != "" {
```

Comment:
Line 576, did you mean tempProvince?

`if tempProvince != ""`

Reply:
fixed